### ISSUE TITLE
Illegal stateexception for suspicious coordinates for node - Europe - GH 4.0 Web Jar

### ISSUE THREAD

**Issue by landliebe** [graphhopper/graphhopper#2461]
**Describe the bug**
Graphhopper is not able to import europe latest pbf

What happens at some point:
java.lang.RuntimeException: Couldn't process file europe-latest.osm.pbf, error: Suspicious coordinates for node 60162619: (36.1718877,36.737518) vs. (36.1701893,36.738058,NaN)

Someone already had that problem in June 2021 with graphhopper 3.0 it seems, but the bug report was closed:
https://discuss.graphhopper.com/t/about-illegalstateexception-for-suspicious-coordinates-for-node/6602/2

**To Reproduce**

Europe latest from about 2 weeks ago and yesterday (16.Nov 2021, aroung 20:00 UTC)
from: **https://download.openstreetmap.fr/extracts/**

**Expected behavior**
That it imports without an Error and bulds the graph cache

**System Information**
graphhopper 4.0 release JAR, Windows 10, OpenJDK 11.0.12

**Screenshots & Logs**

[graphhopper_import_error_2_weeks_old_europe.txt](https://github.com/graphhopper/graphhopper/files/7555315/graphhopper_import_error_2_weeks_old_europe.txt)
[graphhopper_import_error_16_nov_2021.txt](https://github.com/graphhopper/graphhopper/files/7555319/graphhopper_import_error_16_nov_2021.txt)

**Comment by easbar** [comment]
I'll try to reproduce this for: https://download.openstreetmap.fr/extracts/europe-latest.osm.pbf
![image](https://user-images.githubusercontent.com/17603532/142222560-d5852356-94e4-4772-b1a1-deaba35ac088.png)

The 'suspicious' coordinate seems to be right at the border of the Europe extract, near the border between Turkey and Syria: https://graphhopper.com/maps/?point=36.171888%2C36.737518. That could be part of it.

**Comment by easbar** [comment]
Can you please also share the `config.yml` file you used?

**Comment by landliebe** [comment]
Strange thing is:
In both cases it is the Lng/Lat of that point in the Exception message
java.lang.IllegalStateException: Suspicious coordinates for node 59876869: (36.1718877,36.737518) vs. (36.1701893,36.738058,NaN)
java.lang.IllegalStateException: Suspicious coordinates for node 60162619: (36.1718877,36.737518) vs. (36.1701893,36.738058,NaN)

But if we look at the Node IDs:
https://www.openstreetmap.org/node/59876869
https://www.openstreetmap.org/node/60162619

One time on the Falkland Islands on the southern hemisphere and the other time new Hannover/Germany

**Comment by easbar** [comment]
> One time on the Falkland Islands on the southern hemisphere and the other time new Hannover/Germany

Ok this is strange indeed :)

**Comment by landliebe** [comment]
Like the example config but with following changes:  

profiles:
    - name: motorcycle_fastest
      vehicle: car
      weighting: fastest
      
    - name: motorcycle_shortest
      vehicle: car
      weighting: shortest
      
    - name: motorcycle_curvature
      vehicle: motorcycle
      weighting: curvature

  profiles_ch:
    - profile: motorcycle_fastest
    - profile: motorcycle_shortest
    - profile: motorcycle_curvature

**Comment by landliebe** [comment]
ok just finished my test with the original config from:
https://github.com/graphhopper/graphhopper/blob/master/config-example.yml
With the Europe file form yesterday

Again: Exception
Log:
[original_config_exception_europe_16_nov_2021.txt](https://github.com/graphhopper/graphhopper/files/7555624/original_config_exception_europe_16_nov_2021.txt)

**Comment by easbar** [comment]
I can reproduce this, but only for the extract from openstreetmap.fr. Using the extract from Geofabrik everything seems fine.

**Comment by easbar** [comment]
> One time on the Falkland Islands on the southern hemisphere and the other time new Hannover/Germany

No, wait. These are internal GH node IDs, not OSM node IDs. The information that is actually relevant is this:

```
Couldn't properly add edge with osm ids:[1005030478, 4170133500, 1005030422, 4170133499, 8484835379, 1387855317, 7410227416, 7410227427, 186618975, 1005030419, 3170650878, 9126391916, 186618977, 6579545603, 186618979, 1005030376, 9126308678, 186618980, 6579545796, 6579545795, 1005030390, 186618983, 186618984, 6579545781, 6579545810, 6579545769, 7410148598, 6579545764, 1005030470, 6579545733, 3170650958, 6579545837, 1005030371, 186618985, 4170099792, 8111781662, 186618986, 8111781663, 2018655435, 5346676488, 1005030441, 186618987, 4170099870, 186618988, 4170099871, 186618989, 8111841551, 8111841552, 1286376293, 186618992, 4065449517]
```

and this corresponds to this way: https://www.openstreetmap.org/way/113319111, so it is in Syria (very close to the extract borders) indeed. Is there a smaller extract that covers this area? For example Turkey? I could not find one here: https://download.openstreetmap.fr/extracts/

**Comment by landliebe** [comment]
Geofabrik europe import went through successfully over the night. Thanks for the help.

**Comment by easbar** [comment]
I also tried importing the openstreetmap.fr extract using the current master. This also works. So there seems to be some issue that only shows up for this specific extract that was fixed between the 4.0 release and now...

**Comment by akamuza** [comment]
I've had this kind of error two days in a row now for two consecutive PBFs for Ukraine.  GH 4.0, Geofabrik download for Ukraine: https://download.geofabrik.de/europe/ukraine-latest.osm.pbf
The issue occurs with this way: https://www.openstreetmap.org/way/846903409 around 47.813205,29.8186892
The gh error happened just after some user edited the way in OSM two days ago.

Error:
```
com.graphhopper.reader.osm.OSMReader     : Couldn't properly add edge with osm ids:[2212875445, 9514052350, 2212875467, 9514052351, 9514052622, 2212875505, 2212875518, 9514052352, 2212875527, 2212875540, 9514052354, 2212875557, 2212875568, 9514052353, 2212875599, 2212875612, 2212875618, 2212875635, 7049294558, 2212875646, 2212875667, 2212875726, 7049294559, 7049294561, 7049294560, 7049294566, 9514053097, 9514053096, 7049294564, 7049294563, 7049294562, 7049294565, 2212875741, 7010865446, 2212875748, 2212875759, 2212875765, 2212875775, 9514053502, 2212875781, 2212875785, 2212875799, 9514053501, 2212875807, 9514053500, 2212875805, 2212875809, 2212875816, 2212875823, 2212875838, 2212875852, 2212875859, 2212875861, 9514180881, 2212875886, 9514180883, 2212875903, 9514180882, 2212875907, 2212875892, 9514180884, 2212875890, 9514053377, 2212875877, 2212875873, 2212875865, 2212875863, 2212875867, 9514053378, 2212875875, 9514180885, 2212875911, 2212875920, 9514053362, 9514053346, 2212875934, 2212875944, 2212875946, 2212875950, 2212875959, 2212875969, 2212875986, 2212876011, 2212876016, 2212876026, 2212876032, 7051535395, 2212876039, 9514180741, 2212876037, 2212876035, 2212876028, 2212876022, 2212876020, 2212876015, 9514373967, 9514373990, 2212875996, 2212876000, 7051535394, 2212875998, 2212875988, 2212875970, 2212875967, 2212875965, 2212875961, 7051535391, 9514374014, 2212875955, 7051535393, 2212875957, 7051535392, 2212875997, 9514377317, 2212876030, 9514377318, 2212876045, 2212876049, 9514377319, 2212876053, 2212876055, 8091548008, 2212876060, 7051535390, 2212876149, 7051535389, 2212876172, 2212876187, 2212876267, 2212876282, 2212876322, 8159113891, 2212876352, 2212876361, 2212876373, 9514377632, 2212876394, 2212876401, 2212876415, 2212876435, 2212876486, 2212876502, 2212876527, 9514377610, 2212876533, 2212876555, 2212876579, 2212876633, 5415957630, 2212876664, 2212876711, 2212876779, 2212876812, 2212876818, 2212876831, 2212876833, 2212876835, 2850392159, 2212876837, 2212876839, 2212876847, 8060497997, 8060497998, 4743246351, 8060498000, 2212876851, 8060497999, 4743246349]

java.lang.IllegalStateException: Suspicious coordinates for node 2009748: (47.813205,29.8186892) vs. (47.8135259,29.818154,NaN)
```

**Comment by akamuza** [comment]
After some research here is the method that fails:
```java
private void checkCoordinates(int nodeIndex, GHPoint point) {
        final double tolerance = 1.e-6;
        if (Math.abs(nodeAccess.getLat(nodeIndex) - point.getLat()) > tolerance || Math.abs(nodeAccess.getLon(nodeIndex) - point.getLon()) > tolerance)
            throw new IllegalStateException("Suspicious coordinates for node " + nodeIndex + ": (" + nodeAccess.getLat(nodeIndex) + "," + nodeAccess.getLon(nodeIndex) + ") vs. (" + point + ")");
    }
```
Osm way id that fails: 846903409

**Comment by easbar** [comment]
Did you try using master?

**Comment by akamuza** [comment]
In my software I use GH4.0 dependency from maven. 

Here is another smaller file with the same error (34Mb) : https://s3.eu-central-1.amazonaws.com/akcent.pub/osm/odessa_oblast-latest.osm.pbf 
The problem in gh4.0 occurs for the way https://www.openstreetmap.org/way/1012228399

I tried to run tests with this file for master - everything imported fine. But master is already 5.0-snapshot and I'm not sure it is production ready. 
Is there a possibility to get fix for current GH4.0 version?

Besides I didn't manage to run tests for 650Mb Ukraine file due to constant OOM error even with 24Gb heap. Not sure if it should be another issue or just me.

**Comment by easbar** [comment]
> But master is already 5.0-snapshot and I'm not sure it is production ready. Is there a possibility to get fix for current GH4.0 version?

master can usually be considered stable, but I just released `5.0-pre1` as well. 

> Besides I didn't manage to run tests for 650Mb Ukraine file due to constant OOM error even with 24Gb heap. Not sure if it should be another issue or just me.

I just ran the OSM import for `ukraine-220219.osm.pbf` from Geofabrik (one profile: car/fastest) without errors using `-Xmx20g -Xms20g`.

**Comment by easbar** [comment]
Using `ukraine-220219.osm.pbf` from Geofabrik with a car/fastest profile I get this error:

  ```
java.lang.IllegalStateException: Suspicious distance for edge: 934005 2277655-2277656 145.892 vs. 355.78495746511203, difference: -209.89295746511203
  ```

using GH 3.2 and this error:

  ```
java.lang.IllegalStateException: Suspicious distance for edge: 925560 2269210-2269211 145.892 vs. 355.78810307356486, difference: -209.89610307356486
  ```

using GH 4.0. I do not get an error using `ukraine-220101.osm.pbf`.

**Comment by akamuza** [comment]
For Ukraine I ended up splitting that way 846903409 into several segments yesterday and now the problem during import is gone for geofabrik pbf. So for current snapshot I think there is no problems with suspicious coordinates.

But this https://s3.eu-central-1.amazonaws.com/akcent.pub/osm/odessa_oblast-latest.osm.pbf - still has this problem for gh4.0

**Comment by easbar** [comment]
> Edit: Ukraine had this error only during several days: 17-19 Feb I think. I have saved pbf for 19 Feb to replicate the problem in case you need it.

Ok thanks, but I just took it from here: https://download.geofabrik.de/europe/ukraine-220219.osm.pbf

> But this https://s3.eu-central-1.amazonaws.com/akcent.pub/osm/odessa_oblast-latest.osm.pbf - still has this problem for gh4.0

Can you just use GraphHopper 5.0-pre1? It's here: https://repo.maven.apache.org/maven2/com/graphhopper/graphhopper-parent/5.0-pre1/

**Comment by akamuza** [comment]
I guess I have to refactor a bit in order to use 5.0 but sure, I'll give it a try
Thanks.

**Comment by easbar** [comment]
I don't get the error for `moldova-latest.osm.pbf` from Geofabrik. Do you get the error for any most recent OSM area? If so we should try to grab a mini-extract that reproduces the error via the 'Export' feature on openstreetmap.org for example.

**Comment by akamuza** [comment]
The problem is a bit strange. For Ukraine it's gone now as I've modified the problematic way.
But when I pinpointed the location of the problem yesterday, I tried to extract just that area to be able to debug more easily. And to my surprise when I extracted just area with the full way using https://download.bbbike.org/ - I couldn't reproduce the problem.

Then I decided to extract the whole Odessa region - to which that problematic way belongs - I've downloaded it from http://download.openstreetmap.fr/extracts/europe/ukraine/ - and I had the same error but this time in Moldova for way 1012228399 in Ucrainca  town. Because Odessa area contains some border areas of Moldova.

So I tried to export just Ucrainca area using https://download.bbbike.org/ and again - the problem was gone for that pbf. But I've just downloaded http://download.openstreetmap.fr/extracts/europe/ukraine/odessa_oblast-latest.osm.pbf a few minutes ago, and the problem still exists during import on GF4.0: 
```
Suspicious coordinates for node 114341: (46.4298027,29.2784187) vs. (46.4300336,29.2783441,NaN)
```

So, it looks like the problem depends on area that is exported, and only in some cases this glitch occurs.

**Comment by easbar** [comment]
I managed to reproduce the error for this mini OSM file (a small extract around the 'suspicious coordinate' taken from your odessa_oblast file):

https://gist.github.com/easbar/c7d9bd99208ed1f39aa8536e122d4d33

**Comment by easbar** [comment]
I am pretty sure this is somehow related to OSM ways containing missing nodes. This happens we extract an area such that some ways are only partly included. Dealing with this was quite a headache before #2448 so not sure if it is worth fixing this for GH <=4.0.

**Comment by akamuza** [comment]
I've been using graphopper about 5 years, and this error happened only once to me, so I guess it's pretty safe just to ignore this issue for now.

Thanks for feedback on this error.
